pkg/sentry/kernel/mq/mq.go

```diff
@@ -0,0 +1,396 @@
+	// notified when a new message is inserted in the queue.
+	receivers waiter.Queue
+
+	// messages is a list of messages currently in the queue.
+	messages msgList
+
+	// subscriber represents a task registered to receive async notification
+	// from this queue.
+	subscriber *Subscriber
+
+	// messageCount is the number of messages currently in the queue.
+	messageCount int64
+
+	// maxMessageCount is the maximum number of messages that the queue can
+	// hold.
+	maxMessageCount int64
+
+	// maxMessageSize is the maximum size of a message held by the queue.
+	maxMessageSize uint64
+
+	// byteCount is the number of bytes of data in all messages in the queue.
+	byteCount uint64
+}
+
+// View is a view into a message queue. A View provides one of three types of
+// message queues based on options provided to open_mq(2): send and receive
+// queue, send-only queue, and receive-only queue. A send-only queue provides
+// an empty View.Receive, etc.
+//
+// Views should only be used in file descriptions, but not inodes, because we
+// use inodes to retreive the actual queue, and only FDs are responsible for
+// providing user functionality.
+type View interface {
+	// TODO: Add Send and Receive when mq_timedsend(2) and mq_timedreceive(2)
+	// are implemented.
+
+	// Flush checks if the calling process has attached a notification request
+	// to this queue, if yes, then the request is removed, and another process
+	// can attach a request.
+	Flush(ctx context.Context)
+
+	waiter.Waitable
+}
+
+// ReaderWriter provides a send and receive view into a queue.
+type ReaderWriter struct {
+	*Queue
+
+	block bool
+}
+
+// Reader provides a send-only view into a queue.
+type Reader struct {
+	*Queue
+
+	block bool
+}
+
+// Writer provides a receive-only view into a queue.
+type Writer struct {
+	*Queue
+
+	block bool
+}
+
+// NewView creates a new view into a queue and returns it.
+func NewView(q *Queue, rOnly, wOnly, readWrite, block bool) (View, error) {
+	switch {
+	case rOnly:
+		return Reader{Queue: q, block: block}, nil
+	case wOnly:
+		return Writer{Queue: q, block: block}, nil
+	case readWrite:
+		return ReaderWriter{Queue: q, block: block}, nil
+	default:
+		return nil, linuxerr.EINVAL
+	}
+}
+
+// Message holds a message exchanged through a Queue via mq_timedsend(2) and
+// mq_timedreceive(2), and additional info relating to the message.
+//
+// +stateify savable
+type Message struct {
+	msgEntry
+
+	// Text is the message's sent content.
+	Text []byte
+
+	// Size is the message's size in bytes.
+	Size uint64
+
+	// Priority is the message's priority.
+	Priority uint64
+}
+
+// Subscriber represents a task registered for async notification from a Queue.
+//
+// +stateify savable
+type Subscriber struct {
+	// notifyAt is a channel used to deliver async notification to a task when a
+	// new message arrives when the queue is empty. notify should be used with
+	// Task.Block.
+	notifyAt chan struct{}
```

Comment:
Does this work? I don't think channels can be saved by stateify.

Reply:
This is not useful currently, I planned to use it when `mq_notify(2)` is implemented. I removed it from here and added a TODO comment, so we probably should delay discussing it until `mq_notify(2)` is implemented.